File: solutions/typescript/custom-set/1/custom-set.ts
export class CustomSet {
  // Хранилище для уникальных элементов множества
  private elements: unknown[] = [];

  constructor(initial?: unknown[]) {
    // Если передали начальные элементы, добавляем их без повторов
    if (initial) {
      for (const element of initial) {
        this.add(element);
      }
    }
  }

  // Проверяет, пустое ли множество
  empty(): boolean {
    return this.elements.length === 0;
  }

  // Проверяет, есть ли элемент в множестве
  contains(element: unknown): boolean {
    return this.elements.includes(element);
  }

  // Добавляет элемент, если его еще нет в множестве
  add(element: unknown): CustomSet {
    if (!this.contains(element)) {
      this.elements.push(element);
    }
    return this;
  }

  // Проверяет, является ли текущее множество подмножеством другого множества
  // (то есть все ли наши элементы есть в "other")
  subset(other: CustomSet): boolean {
    return this.elements.every((element) => other.contains(element));
  }

  // Проверяет, являются ли множества "непересекающимися"
  // (то есть у них нет ни одного общего элемента)
  disjoint(other: CustomSet): boolean {
    return this.elements.every((element) => !other.contains(element));
  }

  // Проверяет, равны ли множества
  // (они равны, если имеют одинаковую длину и все элементы совпадают)
  eql(other: CustomSet): boolean {
    if (this.elements.length !== other.elements.length) {
      return false;
    }
    return this.subset(other);
  }

  // Объединение множеств: создает новое множество со всеми элементами из обоих
  union(other: CustomSet): CustomSet {
    const newSet = new CustomSet(this.elements);
    for (const element of other.elements) {
      newSet.add(element);
    }
    return newSet;
  }

  // Пересечение множеств: создает новое множество только из общих элементов
  intersection(other: CustomSet): CustomSet {
    const commonElements = this.elements.filter((element) => other.contains(element));
    return new CustomSet(commonElements);
  }

  // Разность множеств: создает новое множество из элементов, которых нет в "other"
  difference(other: CustomSet): CustomSet {
    const uniqueElements = this.elements.filter((element) => !other.contains(element));
    return new CustomSet(uniqueElements);
  }
}

File: solutions/kotlin/custom-set/1/src/main/kotlin/CustomSet.kt
class CustomSet(vararg elements: Int) {

    // Внутреннее хранилище для уникальных элементов
    private val storage = mutableListOf<Int>()

    init {
        // Добавляем только уникальные элементы при создании объекта
        for (element in elements) {
            if (!storage.contains(element)) {
                storage.add(element)
            }
        }
    }

    // Проверяет, пустое ли множество
    fun isEmpty(): Boolean {
        return storage.isEmpty()
    }

    // Проверяет, является ли текущее множество подмножеством другого множества
    fun isSubset(other: CustomSet): Boolean {
        for (element in storage) {
            if (!other.contains(element)) {
                return false
            }
        }
        return true
    }

    // Проверяет, что у множеств нет общих элементов (они не пересекаются)
    fun isDisjoint(other: CustomSet): Boolean {
        for (element in storage) {
            if (other.contains(element)) {
                return false
            }
        }
        return true
    }

    // Проверяет, содержит ли множество конкретное число
    fun contains(other: Int): Boolean {
        return storage.contains(other)
    }

    // Возвращает новое множество, содержащее только общие элементы
    fun intersection(other: CustomSet): CustomSet {
        val commonElements = storage.filter { other.contains(it) }.toIntArray()
        return CustomSet(*commonElements)
    }

    // Добавляет элемент в множество, если его там еще нет
    fun add(other: Int) {
        if (!storage.contains(other)) {
            storage.add(other)
        }
    }

    // Сравнивает два множества на равенство (независимо от порядка элементов)
    override fun equals(other: Any?): Boolean {
        if (other !is CustomSet) return false
        if (this.storage.size != other.storage.size) return false
        return this.isSubset(other)
    }

    // Хэш-код для корректной работы сравнения
    override fun hashCode(): Int {
        return storage.sorted().hashCode()
    }

    // Оператор плюс: объединяет два множества в одно новое
    operator fun plus(other: CustomSet): CustomSet {
        val result = CustomSet(*this.storage.toIntArray())
        for (element in other.storage) {
            result.add(element)
        }
        return result
    }

    // Оператор минус: удаляет из первого множества все элементы, которые есть во втором
    operator fun minus(other: CustomSet): CustomSet {
        val remainingElements = storage.filter { !other.contains(it) }.toIntArray()
        return CustomSet(*remainingElements)
    }
}
